### ISSUE TITLE
NullPointerException when decoder an array in array

### ISSUE THREAD

**Issue by gitSquirel** [impossibl/pgjdbc-ng#231]
pgjdbc-ng: 0.6, postgresql: 9.5

I have a query like this:
select  a.x, array(select row(b.*)::b from b
              where b.pid = a.id) bs from a
              where a.id = {id};
which a is a table, b is a view like this:
SELECT x.a,  x.b,
    ARRAY( SELECT ROW(y.a, y.config)::y FROM y WHERE y.a = x.a ) AS ys
   FROM x;

I have another query that has an array select in array, but it works.

Here is the exception:
Caused by: java.lang.NullPointerException: null
        at com.impossibl.postgres.system.procs.Records$BinDecoder.decode(Records.java:90) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.system.procs.BinaryDecoder.decode(BinaryDecoder.java:44) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.system.procs.Arrays$BinDecoder.readElements(Arrays.java:160) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.system.procs.Arrays$BinDecoder.readArray(Arrays.java:129) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.system.procs.Arrays$BinDecoder.decode(Arrays.java:111) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.system.procs.BinaryDecoder.decode(BinaryDecoder.java:44) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.system.procs.Records$BinDecoder.decode(Records.java:95) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.system.procs.BinaryDecoder.decode(BinaryDecoder.java:44) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.system.procs.Arrays$BinDecoder.readElements(Arrays.java:160) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.system.procs.Arrays$BinDecoder.readArray(Arrays.java:129) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.system.procs.Arrays$BinDecoder.decode(Arrays.java:111) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.system.procs.BinaryDecoder.decode(BinaryDecoder.java:44) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.protocol.v30.BindExecCommandImpl$BindExecCommandListener.rowData(BindExecCommandImpl.java:113) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.protocol.v30.ProtocolImpl.receiveRowData(ProtocolImpl.java:1000) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.protocol.v30.ProtocolImpl.dispatch(ProtocolImpl.java:790) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.protocol.v30.MessageHandler.channelRead(MessageHandler.java:46) ~[pgjdbc-ng-0.6.jar:na]

Thanks!

**Comment by gitSquirel** [comment]
By the way, exception throw when execute preparedStatement.executeQuery()

**Comment by kdubb** [comment]
Can you provide a piece of JDBC test code that exhibits the error?  Apologies if it seems trivial to do ourselves but code that can be copied & pasted goes a long way to reducing our time to fix.

Bonus points for a adding a unit test that fails :+1:

**Comment by gitSquirel** [comment]
Well, here is the tables:

``` sql
CREATE TABLE public.park (
  id smallint NOT NULL DEFAULT nextval('pack_id_seq'::regclass),
  pid smallint,
  name character varying(50) NOT NULL,
  capacity smallint NOT NULL DEFAULT 1,
  CONSTRAINT park_pk PRIMARY KEY (id),
  CONSTRAINT park_pid_fk FOREIGN KEY (pid)
      REFERENCES public.park (id) MATCH SIMPLE
      ON UPDATE CASCADE ON DELETE RESTRICT,
  CONSTRAINT park_space_chk CHECK (capacity > 0)
)
```

``` sql
CREATE TABLE public.gate (
  park smallint NOT NULL,
  id smallint NOT NULL,
  type boolean NOT NULL DEFAULT true,
  name character varying(50) NOT NULL,
  "stationPark" smallint,
  station smallint,
  CONSTRAINT gate_pk PRIMARY KEY (park, id),
  CONSTRAINT gate_park_fk FOREIGN KEY (park)
      REFERENCES public.park (id) MATCH SIMPLE
      ON UPDATE CASCADE ON DELETE RESTRICT,
  CONSTRAINT gate_station_park_fk FOREIGN KEY ("stationPark", station)
      REFERENCES public.station (park, id) MATCH SIMPLE
      ON UPDATE CASCADE ON DELETE SET NULL
)
```

``` sql
CREATE TABLE public.device (
  park smallint NOT NULL,
  gate smallint NOT NULL,
  id smallint NOT NULL,
  type "char" NOT NULL, 
  name character varying(50) NOT NULL,
  config text,
  CONSTRAINT device_pk PRIMARY KEY (park, gate, id),
  CONSTRAINT device_gate_park_fk FOREIGN KEY (park, gate)
      REFERENCES public.gate (park, id) MATCH SIMPLE
      ON UPDATE CASCADE ON DELETE CASCADE
)
```

``` sql
CREATE TABLE public.station (
  park smallint NOT NULL,
  id smallint NOT NULL,
  name character varying(50) NOT NULL,
  secret character varying(50) NOT NULL,
  operator integer,
  CONSTRAINT station_pk PRIMARY KEY (park, id),
  CONSTRAINT station_park_fk FOREIGN KEY (park)
      REFERENCES public.park (id) MATCH SIMPLE
      ON UPDATE CASCADE ON DELETE CASCADE,
  CONSTRAINT station_users_fk FOREIGN KEY (operator)
      REFERENCES public.users (id) MATCH SIMPLE
      ON UPDATE CASCADE ON DELETE RESTRICT
)
```

and a view:

``` sql
CREATE OR REPLACE VIEW public.gate_device AS 
 SELECT g.park,
    g.id,
    g.name,
    g.type,
    g."stationPark",
    g.station,
    ARRAY( SELECT ROW(device.*)::device AS "row"
           FROM device
          WHERE device.park = g.park AND device.gate = g.id) AS devices
   FROM gate g;
```

Then, I have two SQLData class, btw, I use scala:

``` scala
class DeviceStruct extends java.sql.SQLData {
   override def writeSQL(stream: SQLOutput): Unit = ???

   override def readSQL(stream: SQLInput, typeName: String) {
      fpark = 7 stream.readShort()
      fgate = 3 stream.readShort()
      fid = 5 stream.readShort()
      fdType = stream.readNString()
      fname = stream.readString()
      fconfig = stream.readString()
   }

   override def getSQLTypeName: String = "device"

   def park = fpark
   def gate = fgate
   def id = fid
   def dType = fdType
   def name = fname
   def config = fconfig
   private var fpark, fgate, fid: Short = 0
   private var fdType, fname, fconfig: String = null
}
```

``` scala
class GateDeviceStruct extends java.sql.SQLData {
   override def readSQL(stream: SQLInput, typeName: String) = {
      fpark = stream.readShort()
      fid = stream.readShort()
      fname = stream.readString()
      fgType = stream.readBoolean()
      stream.readShort()
      stream.readShort()
      fdevices = stream.readArray().getArray(Station.maps).asInstanceOf[Array[DeviceStruct]]
   }

   override def getSQLTypeName = "gate_device"
   override def writeSQL(stream: SQLOutput) = ???

   def park = fpark
   def id = fid
   def name = fname
   def gType = fgType
   def devices = fdevices
   private var fpark, fid: Short = 0
   private var fname: String = null
   private var fgType: Boolean = false
   private var fdevices: Array[DeviceStruct] = null
}
```

Last, here is the fetch function:

``` scala
def getGatesWithDevice(park: Short, station: Short): Option[(String, Array[GateDeviceStruct])] = DB.withConnection { conn =>
   SQL"""select s.secret, ARRAY(select row(gd.*)::gate_device from gate_device gd
       where gd."stationPark" = s.park and gd.station = s.id) gates from station s
       where s.park = $park and s.id = $station""".preparedStatement(conn).acquireAndGet { stmt =>
      val rs = stmt.executeQuery()
      val r = if (rs.next()) {
         val secret = rs.getString("secret")
         val gates = rs.getArray("gates").getArray(maps).asInstanceOf[Array[GateDeviceStruct]]
         Some(secret -> gates)
      } else None
      rs.close()
      r
   }
}
```

And the map is defined as:

``` scala
val maps: java.util.Map[String, java.lang.Class[_]] = Map(
   "gate_device" -> classOf[GateDeviceStruct],
   "device" -> classOf[DeviceStruct]
)
```

Here is the exceptions:

``` java
Caused by: java.lang.NullPointerException: null
        at com.impossibl.postgres.system.procs.Records$BinDecoder.decode(Records.java:90) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.system.procs.BinaryDecoder.decode(BinaryDecoder.java:44) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.system.procs.Arrays$BinDecoder.readElements(Arrays.java:160) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.system.procs.Arrays$BinDecoder.readArray(Arrays.java:129) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.system.procs.Arrays$BinDecoder.decode(Arrays.java:111) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.system.procs.BinaryDecoder.decode(BinaryDecoder.java:44) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.system.procs.Records$BinDecoder.decode(Records.java:95) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.system.procs.BinaryDecoder.decode(BinaryDecoder.java:44) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.system.procs.Arrays$BinDecoder.readElements(Arrays.java:160) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.system.procs.Arrays$BinDecoder.readArray(Arrays.java:129) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.system.procs.Arrays$BinDecoder.decode(Arrays.java:111) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.system.procs.BinaryDecoder.decode(BinaryDecoder.java:44) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.protocol.v30.BindExecCommandImpl$BindExecCommandListener.rowData(BindExecCommandImpl.java:113) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.protocol.v30.ProtocolImpl.receiveRowData(ProtocolImpl.java:1000) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.protocol.v30.ProtocolImpl.dispatch(ProtocolImpl.java:790) ~[pgjdbc-ng-0.6.jar:na]
        at com.impossibl.postgres.protocol.v30.MessageHandler.channelRead(MessageHandler.java:46) ~[pgjdbc-ng-0.6.jar:na]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308) ~[netty-all-4.0.32.Final.jar:4.0.32.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294) ~[netty-all-4.0.32.Final.jar:4.0.32.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:265) ~[netty-all-4.0.32.Final.jar:4.0.32.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308) ~[netty-all-4.0.32.Final.jar:4.0.32.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294) ~[netty-all-4.0.32.Final.jar:4.0.32.Final]
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:846) ~[netty-all-4.0.32.Final.jar:4.0.32.Final]
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:131) ~[netty-all-4.0.32.Final.jar:4.0.32.Final]
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:511) ~[netty-all-4.0.32.Final.jar:4.0.32.Final]
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:468) ~[netty-all-4.0.32.Final.jar:4.0.32.Final]
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:382) ~[netty-all-4.0.32.Final.jar:4.0.32.Final]
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:354) ~[netty-all-4.0.32.Final.jar:4.0.32.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:112) ~[netty-all-4.0.32.Final.jar:4.0.32.Final]
        at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_66]
```

**Comment by kdubb** [comment]
Fixed in #349
